File: packages/plplot/bindings/octave/demos/x13c.m
## Copyright (C) 1998, 1999, 2000  Joao Cardoso
## Copyright (C) 2004  Rafael Laboissiere
## 
## This program is free software; you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by the
## Free Software Foundation; either version 2 of the License, or (at your
## option) any later version.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## This file is part of plplot_octave.
## It is based on the corresponding demo function of PLplot.
# Does a simple pie chart.

function x13c

  if (!exist("plinit"))
    plplot_stub
  endif

  if(isempty(sprintf("%s",plgdev')))
    plsdev("xwin");
  endif

  text =[
	 "Maurice",
	 "Geoffrey",
	 "Alan",
	 "Rafael",
	 "Vince"
	 ];

  per(1) = 10.;
  per(2) = 32.;
  per(3) = 12.;
  per(4) = 30.;
  per(5) = 16.;

  ## Parse and process command line arguments */

  ## (void) plParseOpts(&argc, argv, PL_PARSE_FULL);

  ## Initialize plplot */
  plinit();

  plenv(0., 10., 0., 10., 1, -2);
  plcol(2);

  theta0 = 0.;
  dthet = 2 * pi / 500;
  for i=0:4
    x = [];
    y = [];
    j = 0;
    x(j+1) = 5.;
    y(j+1) = 5.;
    j=j+1;
    theta1 = theta0 + 2 * pi * per(i+1) / 100.;
    if (i == 4)
      theta1 = 2 * pi;
    endif
    for theta = theta0:dthet:theta1
      x(j+1) = 5 + 3 * cos(theta);
      y(j+1) = 5 + 3 * sin(theta);
      j=j+1;
    endfor
    plcol(i+1);
    plpsty(rem((i + 3), 8) + 1);
    x(j+1:rows(x))=[];
    y(j+1:rows(y))=[];
    plfill(x', y');
    plcol(1);
    plline(x', y');
    just = (theta0 + theta1) / 2.;
    dx = .25 * cos(just); dy = .25 * sin(just);
    if ((just < pi / 2) || (just > 3 * pi / 2) )
      just = 0.;
    else 
      just = 1.;
    endif
    plptex((x((j / 2) +1) + dx), (y((j / 2) +1) + dy), 1.0, 0.0, just, text(i+1,:));
    theta0 = theta - dthet;
  endfor
  plfont(2);
  plschr(0., 1.3);
  plptex(5.0, 9.0, 1.0, 0.0, 0.5, "Percentage of Sales");
  plend1();
endfunction

File: packages/plplot/examples/c/x13c.c
/* $Id: x13c.c 3186 2006-02-15 18:17:33Z slbrow $

	Pie chart demo.
*/

#include "plcdemos.h"

static char *text[] =
{
    "Maurice",
    "Geoffrey",
    "Alan",
    "Rafael",
    "Vince"
};

/*--------------------------------------------------------------------------*\
 * main
 *
 * Does a simple pie chart.
\*--------------------------------------------------------------------------*/

int
main(int argc, char *argv[])
{
    int i, j, dthet, theta0, theta1, theta;
    PLFLT just, dx, dy;
    static PLFLT x[500], y[500], per[5];

    per[0] = 10.;
    per[1] = 32.;
    per[2] = 12.;
    per[3] = 30.;
    per[4] = 16.;

/* Parse and process command line arguments */

    (void) plparseopts(&argc, argv, PL_PARSE_FULL);

/* Initialize plplot */

    plinit();

    plenv(0., 10., 0., 10., 1, -2);
    plcol0(2);
    /* n.b. all theta quantities scaled by 2*pi/500 to be integers to avoid
     * floating point logic problems. */
    theta0 = 0;
    dthet = 1;
    for (i = 0; i <= 4; i++) {
	j = 0;
	x[j] = 5.;
	y[j++] = 5.;
        /* n.b. the theta quantities multiplied by 2*pi/500 afterward so
	 * in fact per is interpreted as a percentage. */
	theta1 = theta0 + 5 * per[i];
	if (i == 4)
	    theta1 = 500;
	for (theta = theta0; theta <= theta1; theta += dthet) {
	    x[j] = 5 + 3 * cos((2.*PI/500.)*theta);
	    y[j++] = 5 + 3 * sin((2.*PI/500.)*theta);
	}
	plcol0(i + 1);
	plpsty((i + 3) % 8 + 1);
	plfill(j, x, y);
	plcol0(1);
	plline(j, x, y);
	just = (2.*PI/500.)*(theta0 + theta1)/2.;
	dx = .25 * cos(just);
	dy = .25 * sin(just);
	if ((theta0 + theta1)  < 250 || (theta0 + theta1) > 750) 
	    just = 0.;
	else 
	    just = 1.;

	plptex((x[j / 2] + dx), (y[j / 2] + dy), 1.0, 0.0, just, text[i]);
	theta0 = theta - dthet;
    }
    plfont(2);
    plschr(0., 1.3);
    plptex(5.0, 9.0, 1.0, 0.0, 0.5, "Percentage of Sales");

/* Don't forget to call PLEND to finish off! */

    plend();
    exit(0);
}
